sql-plugin/src/main/scala/com/nvidia/spark/rapids/CostBasedOptimizer.scala

```diff
@@ -115,25 +118,55 @@ class CostBasedOptimizer(conf: RapidsConf) extends Logging {
     if (totalGpuCost > totalCpuCost) {
       // we have reached a point where we have transitioned onto GPU for part of this
       // plan but with no benefit from doing so, so we want to undo this and go back to CPU
-      if (plan.canThisBeReplaced) {
+      if (plan.canThisBeReplaced && !consumesQueryStage(plan)) {
         // this plan would have been on GPU so we move it and onto CPU and recurse down
         // until we reach a part of the plan that is already on CPU and then stop
         optimizations.append(ReplaceSection(plan, totalCpuCost, totalGpuCost))
-        plan.recursiveCostPreventsRunningOnGpu()
+        recursiveCostPreventsRunningOnGpu(plan)
       }
 
       // reset the costs because this section of the plan was not moved to GPU
       totalGpuCost = totalCpuCost
     }
 
-    if (!plan.canThisBeReplaced) {
+    if (!plan.canThisBeReplaced || consumesQueryStage(plan)) {
       // reset the costs because this section of the plan was not moved to GPU
       totalGpuCost = totalCpuCost
     }
 
     (totalCpuCost, totalGpuCost)
   }
 
+  /**
+   * Recursively force a section of the plan back onto CPU, stopping once a plan
+   * is reached that is already on CPU or cannot be moved back onto CPU.
+   */
+  final def recursiveCostPreventsRunningOnGpu(plan: SparkPlanMeta[_]): Unit = {
+    if (plan.canThisBeReplaced && !consumesQueryStage(plan)) {
+      plan.costPreventsRunningOnGpu()
+      plan.childPlans.foreach(recursiveCostPreventsRunningOnGpu)
+    }
+  }
+
+  /**
+   * Determines whether the specified plan will read from a query stage.
+   */
+  private def consumesQueryStage(plan: SparkPlanMeta[_]): Boolean = {
+    if (SQLConf.get.adaptiveExecutionEnabled) {
```

Comment:
nit: if / else looks superficial:

```suggestion
    SQLConf.get.adaptiveExecutionEnabled && plan.wrapped match {
      case _: CustomShuffleReaderExec
             | _: ShuffledHashJoinExec
             | _: BroadcastHashJoinExec
             | _: BroadcastNestedLoopJoinExec => true
         case _ => false
    }
```

Reply:
Thanks. Updated.